core/xmp_parser.cpp

```diff
@@ -0,0 +1,137 @@
+#include "xmp_parser.h"
+#include "subsurface-string.h"
+#include "dive.h" // TODO: monster-include file only for utc_mktime.
+
+#include <libxml/parser.h>
+#include <libxml/tree.h>
+#include <cctype>
+
+static timestamp_t parse_xmp_date(const char *date)
+{
+	// Format: "yyyy-mm-dd[Thh:mm[:ss[.ms]][-05:00]]"
+	int year, month, day;
+	if (sscanf(date, "%d-%d-%d", &year, &month, &day) != 3)
+		return 0;
+
+	int hours = 0, minutes = 0, seconds = 0, milliseconds = 0;
+	int timezone = 0;
+
+	// Check for time part
+	if ((date = strchr(date, 'T')) != nullptr) {
+		++date; // Skip 'T'
+		if (sscanf(date, "%d:%d:%d.%d", &hours, &minutes, &seconds, &milliseconds) < 2)
+			return 0;
+
+		// Check for timezone part. Note that we simply ignore 'Z' as that
+		// means no time zone
+		while (*date && *date != '+' && *date != '-')
+			++date;
+		if (*date) {
+			int sign = *date == '+' ? 1 : -1;
+			int timezone_hours, timezone_minutes;
+			++date;
+			if (sscanf(date, "%d:%d", &timezone_hours, &timezone_minutes) != 2)
+				return 0;
+			timezone = sign * (timezone_hours * 60 + timezone_minutes) * 60;
+		}
+	}
+
+	// Round to seconds, since our timestamps are in seconds
+	if (milliseconds >= 500)
+		seconds += 1;
+
+	struct tm tm = { 0 };
+	tm.tm_year = year - 1900;
+	tm.tm_mon = month - 1;
+	tm.tm_mday = day;
+	tm.tm_hour = hours;
+	tm.tm_min = minutes;
+	tm.tm_sec = seconds;
+
+	timestamp_t res = utc_mktime(&tm);
+	res += timezone;
+
+	return res;
+}
+
+static timestamp_t extract_timestamp_from_attributes(const xmlNode *node)
+{
+	for (const xmlAttr *p = node->properties; p; p = p->next) {
+		const xmlChar *ns = p->ns ? p->ns->prefix : nullptr;
+
+		// Check for xmp::CreateDate property
+		if (!strcmp((const char *)ns, "xmp") && !strcmp((const char *)p->name, "CreateDate")) {
+			// We only support a single property value
+			if (!p->children || !p->children->content)
+				return 0;
+			const char *date = (const char *)p->children->content;
+			return parse_xmp_date(date);
+		}
+	}
+	return 0;
+}
+
+static timestamp_t extract_timestamp(const xmlNode *node)
+{
+	// We use a private stack, so that we can return in one go without
+	// having to unwind the call-stack. We only recurse to a fixed depth,
+	// since the data we are interested in are at a shallow depth.
+	// This can be increased on demand.
+	static const int max_recursion_depth = 16;
```

Comment:
interesting choice. I assume that made it easier? Faster? ??? i.e. what drove the decision to do this with a fixed stack?
(I'm not saying this is *wrong*, I just want to understand it better)

Reply:
> i.e. what drove the decision to do this with a fixed stack?

Laziness. Always suspect laziness first. :-P
It means that we can simple `return` once we have the datum that we're interested in and don't have to check at higher levels if the lower levels found what they're after.

The interface will have to be adapted anyway once we extract more than one datum, so my reasoning is that it's not worth doing this correctly now.

Reply:
In case you ask why it's fixed depth: at first I had an automatically growing `std::vector<>`, but then changed it to a C-array, because I figured it would be easier to stomach. :)